src/macaron/parsers/bashparser.py

```diff
@@ -115,78 +147,109 @@ def parse(bash_content: str, macaron_path: str = "") -> dict:
     ------
-    BashCommands
-        The parsed bash script objects.
+    CallGraphError
+        When unable to create a bash node.
     """
-    if not macaron_path:
-        macaron_path = global_config.macaron_path
-
-    parsed_parent = parse(bash_content)
-    caller_commands = parsed_parent.get("commands", [])
-    if caller_commands:
-        yield BashCommands(
-            caller_path=ci_file,
-            CI_path=ci_file,
-            CI_type=ci_type,
-            commands=caller_commands,
-            job_name=job_name,
-            step_name=step_name,
-        )
+    if recursion_depth > defaults.getint("bashparser", "recursion_depth", fallback=3):
+        raise CallGraphError(f"The analysis has reached maximum recursion depth {recursion_depth} at {source_path}.")
+    parsed_content = {}
+    working_dir = None
+    match node_type:
+        case BashScriptType.INLINE:
+            if parsed_obj is None:
+                raise CallGraphError(f"Unable to find the parsed AST for the CI step at {source_path}.")
+            step_exec = validate_step(parsed_obj)
+            if step_exec is None:
+                raise CallGraphError(f"Unable to validate parsed AST for the CI step at {source_path}.")
+
+            working_dir = step_exec.get("WorkingDirectory")
+            run_script = validate_run_step(parsed_obj)
+            if run_script is None:
+                raise CallGraphError(f"Invalid run step at {source_path}.")
+            parsed_content = parse(run_script, macaron_path=macaron_path)
+        case BashScriptType.FILE:
+            parsed_content = parse_file(source_path, macaron_path=macaron_path)
+    bash_node = BashNode(
+        name,
+        node_type,
+        source_path,
+        parsed_step_obj=parsed_obj,
+        parsed_bash_obj=parsed_content,
+        node_id=node_id,
+        caller=caller,
+    )
+    caller_commands = parsed_content.get("commands", [])
 
     # Parse the bash script files called from the current script.
-    if recursive and repo_path:
+    if caller_commands and repo_path:
         for cmd in caller_commands:
             # Parse the scripts that end with `.sh`.
-            # We only parse recursively at depth 1, so don't set the recursive argument in parse_file().
             # TODO: parse Makefiles for bash commands.
-            if cmd[0] and cmd[0].endswith(".sh") and os.path.exists(os.path.join(repo_path, cmd[0])):
-                callee_commands = parse_file(os.path.join(repo_path, cmd[0])).get("commands", [])
-                if not callee_commands:
-                    continue
-
-                yield BashCommands(
-                    caller_path=os.path.join(working_dir, cmd[0]),
-                    CI_path=ci_file,
-                    CI_type=ci_type,
-                    commands=callee_commands,
-                    job_name=job_name,
-                    step_name=step_name,
-                )
+            if not cmd[0] or not cmd[0].endswith(".sh"):
```

Comment:
I believe this line doesn't raise any exception as long as `cmd` is not an empty list, or else an IndexError will be raised when we try to obtain an indexed element in the list `cmd[0]`. 
I wonder what is the purpose of `cmd[0]`? Is that for checking whether the first element in `cmd` is not an empty string, or is it for checking whether `cmd` is not an empty list :thinking:

Reply:
It is for checking if `cmd[0]` is `None`.

Reply:
Then we have made the assumption that `cmd` is never an empty list here right? If that's the case then we don't need to care about the `IndexError` as I mentioned above.

Reply:
I should have checked if `cmd` is an empty list. I'll add that.

Reply:
See commit [155d65a](https://github.com/oracle/macaron/pull/673/commits/155d65a8c07cb2b6c1c4932bcf43d62d602e82d5)